### PR TITLE
ci: bake appveyor images automatically, run sync on depshash change

### PR DESCRIPTION
**Pull request by VerteDinde**
#### Description of Change

This PR improves our current Windows automation in two ways:
1. Automatically bakes AppVeyor images when the version of Chromium is updated in the base branch (main, 21-x-y, etc).
2. Determines whether or not the depshash of the PR in testing is different than the latest baked image - if it is the same, we run `gclient runhooks` rather than a full `gclient sync`, saving time.

In order to accomplish #1 and keep each branch relatively up to date with the latest baked image, this PR also adds a daily cron job/GitHub Action that checks whether or not the Chromium version in main has changed, kicks off a new image bake and automatically opens a PR updating `appveyor.yml` to the newest bake image.

Merge Blockers:
- [x] The GH Action is working as expected in a fork, but not in `e/e` - this may be because the token used for Actions needs to be scoped to open PRs [_small_]
- [x] ~`win-ia32` is failing "Test" with a node-gyp error for missing symbols. This is not present on x64, woa or the current Azure image, likely due to the updated winSDK SDK.18362 to SDK.20348. [_medium_]~
- [x] ~`x64-release` is failing on `dump_syms.py`, specifically the `subprocess` call on L42. Release jobs are also kicked off on `libcc-20` - we need to either 1) bake a parallel job for that cloud, 2) clone the existing baked image to that cloud, or 3) move release jobs to `electron-16-core`. [_medium_]~
- [x] ~The current Azure image is being built on the new Azure account - migrate that back to the old Azure account on `electron-16-core` before merge [_small_]~

Potential Future Improvement:

- [ ] Currently, the new baked imaged is only kicked off when a new Chromium version is merged. In the future, we could also create a new image whenever the depshash changes (i.e. a patch is changed, etc), to keep builds fast.
- [ ] Even with the pre-build and removed sync, building Electron symbols (specifically `ninja -C out/Default electron:electron_app` can take ~15 minutes. Improving that build time would lead to much faster build jobs.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: none

### CODE REVIEW

File: appveyor-bake.yml [hunk 1]
```diff
@@ -0,0 +1,93 @@
+# The config is used to bake appveyor images, not for running CI jobs.
+# The config expects the following environment variables to be set:
+#  - "GN_CONFIG" Build type. One of {'testing', 'release'}.
+#  - "GN_EXTRA_ARGS" Additional gn arguments for a build config,
+#      e.g. 'target_cpu="x86"' to build for a 32bit platform.
+#      https://gn.googlesource.com/gn/+/master/docs/reference.md#target_cpu
+#      Don't forget to set up "NPM_CONFIG_ARCH" and "TARGET_ARCH" accordningly
+#      if you pass a custom value for 'target_cpu'.
```
Comment (jkleinsc) [inline appveyor-bake.yml:8]: These environment variables are unused and should be removed from the comment

File: appveyor-bake.yml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  GIT_CACHE_PATH: C:\Users\appveyor\libcc_cache
+  ELECTRON_OUT_DIR: Default
+  ELECTRON_ENABLE_STACK_DUMPING: 1
+  MOCHA_REPORTER: mocha-multi-reporters
+  MOCHA_MULTI_REPORTERS: mocha-appveyor-reporter, tap
+  GOMA_FALLBACK_ON_AUTH_FAILURE: true
+  DEPOT_TOOLS_WIN_TOOLCHAIN: 0
+  PYTHONIOENCODING: UTF-8
+build_script:
+  - ps: echo "Building $env:GN_CONFIG build"
```
Comment (jkleinsc) [inline appveyor-bake.yml:26]: This line should be removed.

File: appveyor-bake.yml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  - ps: >-
+      if (Test-Path 'env:RAW_GOMA_AUTH') {
+        $goma_login = python $env:LOCAL_GOMA_DIR\goma_auth.py info
+        if ($goma_login -eq 'Login as Fermi Planck') {
+          Write-warning "Goma authentication is correct";
+        } else {
+          Write-warning "WARNING!!!!!! Goma authentication is incorrect; please update Goma auth token.";
+          $host.SetShouldExit(1)
+        }
+      }
```
Comment (jkleinsc) [inline appveyor-bake.yml:59]: We don't need to setup Goma/build tools in order to bake an image.

File: appveyor.yml [hunk 4]
```diff
@@ -50,163 +51,138 @@ matrix:
...
-              $env:SAVE_GCLIENT_SRC="true"
+            if (Test-Path -Path $zipfile) {
+              # file exists, unzip and then gclient sync
+              7z x -y $zipfile -mmt=14 -aoa
+              if (-not (Test-Path -Path "src\buildtools")) {
+                if ($env:TARGET_ARCH -ne 'ia32') {
+                  # only save on x64/woa to avoid contention saving
+                  $env:SAVE_GCLIENT_SRC="true"
+                }
+              }
```
Comment (jkleinsc) [inline appveyor.yml:139]: This should be removed as this was just used for the old caching strategy.  This if block just needs to set `$env:RUN_GCLIENT_SYNC="true"`

File: script/doc-only-change.js [hunk 5]
```diff
@@ -23,7 +23,7 @@ async function checkIfDocOnlyChange () {
             pullRequestNumber = prsForBranch.data[0].number;
           } else {
             // If there are 0 PRs or more than one PR on a branch, just assume that this is more than a doc change
-            process.exit(1);
+            process.exit(0);
```
Comment (jkleinsc) [inline script/doc-only-change.js:26]: This change should not be made.  The script should exit with an exit code of 1 if is not a doc change.

**Review by jkleinsc — CHANGES_REQUESTED** [review]
Great start here!

File: appveyor-bake.yml (hunk 2 above)
Comment (jkleinsc) [reply appveyor-bake.yml:26]: The line to remove is `- ps: echo "Building $env:GN_CONFIG build"`

File: appveyor-bake.yml (hunk 3 above)
Comment (VerteDinde) [reply appveyor-bake.yml]: Oh awesome - removed all Goma & build tools code!

File: script/doc-only-change.js (hunk 5 above)
Comment (VerteDinde) [reply script/doc-only-change.js]: Sorted, `exit 0` stops this 😄

File: appveyor.yml [hunk 6]
```diff
@@ -300,6 +263,10 @@ for:
           New-Item .\out\Default\gen\node_headers\Release -Type directory
           Copy-Item -path .\out\Default\electron.lib -destination .\out\Default\gen\node_headers\Release\node.lib
       - cd electron
+      - ps: >-
+          if ($env:TARGET_ARCH -eq 'ia32') {
+            $env:npm_config_arch = "ia32"
+          }
```
Comment (dsanders11) [inline appveyor.yml:269]: Could we add a small note that this is a workaround to an issue setting that environment variable in the AppVeyor config? Good to note that this shouldn't be necessary, but is for some reason.

File: appveyor-bake.yml [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  - update_depot_tools.bat
+  - ps: Move-Item $env:APPVEYOR_BUILD_FOLDER -Destination src\electron
+  - src\electron\script\setup-win-for-dev.bat
+  - ps: >-
+      if (Test-Path 'env:RAW_GOMA_AUTH') {
+        $env:GOMA_OAUTH2_CONFIG_FILE = "$pwd\.goma_oauth2_config"
+        $env:RAW_GOMA_AUTH | Set-Content $env:GOMA_OAUTH2_CONFIG_FILE
+      }
+  - ps: >-
+      $env:NINJA_STATUS="[%r processes, %f/%t @ %o/s : %es] "
```
Comment (jkleinsc) [inline appveyor-bake.yml:36]: I don't think these steps are needed for a bake.  They are used for Goma/building/.

File: appveyor.yml [hunk 8]
```diff
@@ -50,164 +51,128 @@ matrix:
...
+            $env:NINJA_STATUS="[%r processes, %f/%t @ %o/s : %es] "
+          }
+      - gclient config --name "src\electron" --unmanaged %GCLIENT_EXTRA_ARGS% "https://github.com/electron/electron"
+      # Patches are applied in the image bake. Check depshash to see if patches have changed. 
+      - ps: $env:RUN_GCLIENT_SYNC="false"
+      - ps: $depshash_baked = Get-Content .\src\.depshash -Raw
+      - ps: cd src\electron
+      - ps: node script\generate-deps-hash.js
+      - ps: $depshash = Get-Content .\.depshash -Raw
+      - ps: $zipfile = "Z:\$depshash.7z"
```
Comment (jkleinsc) [inline appveyor.yml:126]: This line is not needed

**Review by jkleinsc — COMMENTED** [review]
LGTM, just needs some minor cleanup.

Also, comparing to what's running in main, this should save us around ~20 minutes per build!

**Comment by VerteDinde** [comment]
Note: Tested fork PRs with the new images to make sure the build/tests would run properly ([see this PR](https://github.com/electron/electron/pull/36380)). Both the x64 and ia32 jobs ran as expected - I added one environment variable to the ia32 job for `npm_config_target_arch`. I think this is ready for final review & merge!

**Review by jkleinsc — APPROVED** [review]

**Pull request merged**